### ISSUE TITLE
[BUG] Add information for airgapped environments about upgrade process of Security prebuilt detection rules

### ISSUE THREAD

**Issue by 111andre111** [elastic/security-docs#4652]
### Documentation links

https://www.elastic.co/guide/en/elastic-stack/current/air-gapped-install.html
https://www.elastic.co/guide/en/fleet/current/air-gapped.html
https://www.elastic.co/guide/en/security/current/offline-endpoint.html

### Description

Looking to the documentation links we have information about how to set up in an airgapped environment Fleet, Endpoint and other artifacts.

However we should care as well about a detailed guide for maintaining out of band detection rules in kibana:
https://www.elastic.co/guide/en/security/current/prebuilt-rules-downloadable-updates.html
which is coming afaik from this integration:
https://docs.elastic.co/integrations/security_detection_engine

It looks as well like this documentation page is not up to date:
https://www.elastic.co/guide/en/security/current/prebuilt-rules-changelog.html

### Which documentation set(s) does this bug apply to?

ESS and serverless

### Release version

N/A

### Testing environment

on premise installation in airgapped environments.
Maybe might make sense as well to think about ECE and ECK installations.

**Comment by florent-leborgne** [comment]
@approksiu Hi! I've read the issue and am not sure I understand fully the context & request. Can you provide more details? Or point us to a team/someone who can help feed the correct information to add to the docs?

**Comment by approksiu** [comment]
cc @111andre111 @jkelas

**Comment by jkelas** [comment]
If I understand this correctly, @111andre111  would like to document the learnings that we made while working on this ticket https://github.com/elastic/kibana/issues/181808
Am I correct @111andre111 ?
I am thinking about documenting the fact that with the `xpack.fleet.isAirGapped` set to true, Kibana doesn't try to contact EPR, and correctly uses the prepackaged security detection rules, if available (e.g. in the container or in the correct path on the file system).

**Comment by 111andre111** [comment]
This issue here is a slightly other one. https://github.com/elastic/kibana/issues/181808 tries to define what to do when Airgapped and using no EPR.

This issue here is more about Airgapped and using EPR and how to keep Rules artifacts up to date.
This is now the more interesting in the context of the now available rules diff feature
https://www.elastic.co/docs/solutions/security/detect-and-alert/prebuilt-rules-update-modified-unmodified

Hope that helps.

**Comment by jkelas** [comment]
FYI @banderror

**Comment by approksiu** [comment]
I can work on this with @nastasha-solomon.
@nastasha-solomon what do you think if we add a section on this page https://www.elastic.co/docs/solutions/security/detect-and-alert/install-manage-elastic-prebuilt-rules about airgapped environments. We can mention that updates for the rules are shipped via detection-engine integration, so to get the biweekly updates you need to update the EPR https://www.elastic.co/docs/deploy-manage/deploy/self-managed/air-gapped-install#air-gapped-elastic-package-registry, the rest of the rules functionality will remain the same.

**Comment by banderror** [comment]
Hey @approksiu, before we document I think we should test it. Let's chat offline.

**Comment by banderror** [comment]
Hey @pborgonovi, since you worked on testing of https://github.com/elastic/kibana/issues/181808 and are familiar with air-gapped environments, could you please help us with testing here as well?

We want to know if users in air-gapped environments are able to consume OOTB prebuilt rule updates without upgrading their whole stack to the new version. Presumingly, this should be possible for those users whose air-gapped environment contains an air-gapped EPR deployed to it. In other words, for those who run both Kibana and EPR within the air-gapped env. The assumption is: if the user upgrades their air-gapped EPR to the new version, they will get prebuilt rule updates in Kibana.

We want to test it and make sure this works before we document it in the Security documentation.

EPR docs: https://www.elastic.co/docs/deploy-manage/deploy/self-managed/air-gapped-install#air-gapped-elastic-package-registry

**Comment by pborgonovi** [comment]
Hi @banderror 
Sure. I'll work on executing this flow and share the steps + results.

**Comment by deatonlm** [comment]
We are using an air-gapped environment and are interested in this issue.  We are just now having discussions about how we will do rule updates in the future.  Can you please elaborate on what you mean by "OOTB prebuilt rule updates".  We are running the EPR within the same air-gapped environment and on the same server.

**Comment by pborgonovi** [comment]
Hey @banderror 

I was able to validate that prebuilt rules updates are available by upgrading the EPR version, without requiring a full stack upgrade.

The following steps outline the process I used to conduct this test:

#### Objective:
Validate that in a fully air-gapped environment, Kibana can:
- Initially offer only the prebundled detection rules included in the build,
- Successfully install these rules,
- Later detect new rules and rule updates when a newer package version is made available via a local Elastic Package Registry (EPR).

#### Pre-conditions:
- No internet access at any point (completely offline environment).

#### Steps:

1. Initial environment setup without EPR
- Downloaded Elasticsearch and Kibana images version 8.18.4
- Started the Elasticsearch and Kibana Docker containers using `docker-compose`
- Ensured `xpack.fleet.isAirGapped: true` was set in the `kibana.yml` configuration from the beginning
- Did **not** configure `xpack.fleet.registryUrl` at this point

2. Install prebundled detection rules
- Accessed Kibana after the stack finished booting
- Navigated to **Security → Rules**
- Confirmed that only prebundled rules were listed (embedded in the build)
- Selected all available rules and installed them via the UI (a total of 1361 rules)

<img width="1702" height="935" alt="Image" src="https://github.com/user-attachments/assets/3e3a0bf2-1a54-4233-8aab-76669def8e5f" />

<img width="1694" height="916" alt="Image" src="https://github.com/user-attachments/assets/07a42f89-f472-459f-b764-1b35df194158" />


3. Prepare the local EPR with production version 9.1.2:
- Manually downloaded the `security_detection_engine` package version `9.1.2` in `.zip` format
- Extracted the package contents locally
- Placed the extracted contents into the correct structure expected by the local EPR:
  - Directory used: `packages/package-storage/security_detection_engine/9.1.2/`
  - Ensured the directory contained all required files (`manifest.yml`, `docs/`, `img/`, etc.)
- Updated the `docker-compose.yml` file to include the new EPR container with appropriate networking and volume mounts:


```
  epr-new-prod:
    image: docker.elastic.co/package-registry/distribution:production
    container_name: epr-new-prod
    ports:
      - "8081:8080"
    volumes:
      - ./packages/package-storage:/packages/package-storage
    networks:
      elastic818:
        ipv4_address: 172.28.0.101
```
- Updated the kibana.yml file to configure the new EPR:

```
xpack.fleet.registryUrl: "http://epr-new-prod:8080"
xpack.fleet.packages:
  - name: security_detection_engine    
    version: 9.1.2
```

4. Restarted Elasticsearch, Kibana and the new EPR containers.
5. Validated the EPR availability:
- Verified that the EPR is serving the package correctly by sending a GET request to the following URL from within the Docker network:

`http://<epr-container-ip>:8080/package/security_detection_engine/9.1.2`

- Confirmed that a valid JSON response was returned, indicating that the EPR recognized and was serving the package version 9.1.2 correctly.

5. Install the new package in Kibana:
- Manually installed the package in Kibana sending a POST request to Fleet API `/api/fleet/epm/packages/security_detection_engine/9.1.2`

6. Final validation:
- Launched Kibana -> Security -> Rules -> Detection Rules
- Confirmed that:
   - New rules from package version 9.1.2 are now shown as available to install.
   - Some previously installed rules now have updates available.

<img width="3388" height="2118" alt="Image" src="https://github.com/user-attachments/assets/9646b95a-8e67-4891-91a6-c73ce277edaf" />

<img width="1699" height="1024" alt="Image" src="https://github.com/user-attachments/assets/9df84373-702d-471d-9527-408af7938dbc" />

<img width="1684" height="917" alt="Image" src="https://github.com/user-attachments/assets/d9ead536-910f-4a74-9c2d-1b0ac093ceac" />

#### Summary:

Users are able to upgrade prebuilt rules and install new available rules by upgrading EPR without requiring a full stack upgrade.

**Comment by banderror** [comment]
Thanks @pborgonovi, that looks very promising. I do have a few concerns about this testing though.

First of all, I think we shouldn't assume that a user will be sitting on Kibana without an EPR and then adding EPR to their environment. That doesn't sound like what the majority of users will be doing who are interested in OOTB rule updates. The base case we should test is upgrading an `env v1` to `env v2`, where:
- `env v1` contains Kibana of version X and EPR of version Y
- `env v2` contains Kibana of version X and EPR of version Y+1

I'd also like to clarify a couple things about the configuration.

Doesn't the `docker.elastic.co/package-registry/distribution:production` image already contain all the Fleet packages, including all the versions of the `security_detection_engine` package? Based on the [docs](https://www.elastic.co/docs/reference/fleet/air-gapped#air-gapped-diy-epr):

> docker.elastic.co/package-registry/distribution:production - Packages available in the production registry (https://epr.elastic.co/). Note that this image is updated every time a new version of a package gets published.

If true, users won't need to be downloading, unzipping, and mounting the package.

Also, is it necessary to specify `xpack.fleet.packages` in the kibana config? According to the [docs](https://www.elastic.co/docs/reference/kibana/configuration-reference/fleet-settings#_preconfiguration_settings_for_advanced_use_cases), it's a "_list of integrations that are installed when the Fleet app starts up for the first time_". It shouldn't be necessary to install this package on start.

Can we please re-test it according to this?

**Comment by banderror** [comment]
> Can you please elaborate on what you mean by "OOTB prebuilt rule updates". We are running the EPR within the same air-gapped environment and on the same server.

Hi @deatonlm, sure. "OOTB" stands for "Out Of The Band" and means prebuilt rule updates that are being released more frequently than the Elastic stack release cadence. In other words, you can keep your stack and Kibana at the same version, let's say `v9.1.0`, and still continue getting updates for prebuilt rules from Elastic.

**Comment by deatonlm** [comment]
> > Can you please elaborate on what you mean by "OOTB prebuilt rule updates". We are running the EPR within the same air-gapped environment and on the same server.
> 
> Hi [@deatonlm](https://github.com/deatonlm), sure. "OOTB" stands for "Out Of The Band" and means prebuilt rule updates that are being released more frequently than the Elastic stack release cadence. In other words, you can keep your stack and Kibana at the same version, let's say `v9.1.0`, and still continue getting updates for prebuilt rules from Elastic.

Thanks for the reply.  Also, just to let you know, sometimes docker containers are not allowed in air-gapped environments.  We are running the EPR outside of docker under nginx.  We have stripped it down to contain only the version packages that we need in order to reduce its size.  I believe it was over 17GB if I remembering correctly, but we have it down to 2 GB.  Sometimes in air-gapped environments, space is it a premium.

**Comment by pborgonovi** [comment]
Hey @banderror 

I've redone the tests by working with 2 EPR versions: 1 older version and 1 newest version. Find below the detailed steps:

#### Steps:

1. Pulled ES+Kibana version 8.18.4 images
2. Pulled 2 versions of EPR:
    - docker.elastic.co/package-registry/distribution:8.17.0
    - docker.elastic.co/package-registry/distribution:production
3. Disconnected machine from internet access
4. Configured `docker-compose.yml `with both `epr-old-817` and `epr-new-prod`

```
epr-old-817:
   image: docker.elastic.co/package-registry/distribution:8.17.0
   container_name: epr-old-817
   command: ["--log-level=debug"]
   ports:
     - "8080:8080"
   networks:
     elastic818:
       ipv4_address: 172.28.0.100


 epr-new-prod:
   image: docker.elastic.co/package-registry/distribution:production
   container_name: epr-new-prod
   ports:
     - "8081:8080"
   networks:
     elastic818:
       ipv4_address: 172.28.0.101
```

5. Set `kibana.yml` with `xpack.fleet.isAirGapped: true`
6. Initially set `kibana.yml` to use `epr-old-817`:
`xpack.fleet.registryUrl: "http://epr-old-817:8080"`
7. Started elasticsearch, kibana and epr-old-817 containers
8. Containers started and packages were loaded, including `security_detection_engine` package: [epr-old-817.docx](https://github.com/user-attachments/files/21415905/epr-old-817.docx)
9. Launched Kibana -> Security -> Rules -> Detection Rules
10. 1361 prebuilt rules were loaded and available to be installed
11. Installed all available prebuilt rules

https://github.com/user-attachments/assets/8cd570f5-85e4-4eea-9045-32b52c4caced

12. Stopped all containers
13. Updated `kibana.yml` to use `epr-new-prod`
`xpack.fleet.registryUrl: "http://epr-new-prod:8080"`
14. Started elasticsearch, kibana and epr-new-prod containers
15. Containers started and packages were loaded, including `security_detection_engine` package: [epr-new-prod.docx](https://github.com/user-attachments/files/21416114/epr-new-prod.docx)
16. Launched Kibana -> Security -> Rules -> Detection Rules
     - 8 new rules available for installation
     - 37 rule updates

<img width="1702" height="950" alt="Image" src="https://github.com/user-attachments/assets/5044fa61-7422-4cd8-b29d-d203bbff4631" />

https://github.com/user-attachments/assets/e55ecfce-d421-48d8-a844-7efe1a55e2c4


#### Summary:

This flow works as expected — updating the EPR alone is sufficient to:
- Trigger updates for existing prebuilt rules, and
- Make new prebuilt rules from the updated package available for installation.

This proves that users can receive rule updates and additions by simply replacing the local EPR with a newer package set, without requiring a full stack upgrade.

**Comment by pborgonovi** [comment]
> > > Can you please elaborate on what you mean by "OOTB prebuilt rule updates". We are running the EPR within the same air-gapped environment and on the same server.
> > 
> > 
> > Hi [@deatonlm](https://github.com/deatonlm), sure. "OOTB" stands for "Out Of The Band" and means prebuilt rule updates that are being released more frequently than the Elastic stack release cadence. In other words, you can keep your stack and Kibana at the same version, let's say `v9.1.0`, and still continue getting updates for prebuilt rules from Elastic.
> 
> Thanks for the reply. Also, just to let you know, sometimes docker containers are not allowed in air-gapped environments. We are running the EPR outside of docker under nginx. We have stripped it down to contain only the version packages that we need in order to reduce its size. I believe it was over 17GB if I remembering correctly, but we have it down to 2 GB. Sometimes in air-gapped environments, space is it a premium.

@banderror I may be mistaken here, please correct me if I'm wrong - But per my tests I believe that even if running the EPR outside of Docker  it will continue to work as expected.

That’s because what matters is that the Kibana instance is able to reach the EPR endpoint and that the updated package files are available at that location.

As long as the custom EPR setup serves the updated versions of the packages:
- Kibana will detect the new versions automatically.
- Prebuilt rules from updated security_detection_engine packages will show as updatable.
- New rules introduced in newer package versions will appear as installable.

So prebuilt rules updates would be available — even with EPR hosted outside Docker — as long as the package content served includes the correct updates.

The only caveat is: if not using the Docker image (which comes with the packages embedded), the user would need to manually download the .zip files of the desired packages and place them in the package-storage directory (the [additional steps](https://github.com/elastic/security-docs/issues/4652#:~:text=If%20true%2C%20users%20won%27t%20need%20to%20be%20downloading%2C%20unzipping%2C%20and%20mounting%20the%20package.) which had been performed earlier).

**Comment by banderror** [comment]
@deatonlm I appreciate the additional context, thank you.

@pborgonovi That's correct. From the Kibana perspective, it doesn't matter _how_ an instance of EPR is run: under Docker, Nginx, or anyhow else. The two things that matter:

- EPR should be available to Kibana
- EPR should be serving the `security_detection_engine` package with prebuilt rules

> The only caveat is: if not using the Docker image (which comes with the packages embedded), the user would need to manually download the .zip files of the desired packages and place them in the package-storage directory (the https://github.com/elastic/security-docs/issues/4652#:~:text=If%20true%2C%20users%20won%27t%20need%20to%20be%20downloading%2C%20unzipping%2C%20and%20mounting%20the%20package. which had been performed earlier).

Yes, this makes sense.

- If a user uses the production Docker image of EPR in an air-gapped environment (for example, `docker.elastic.co/package-registry/distribution:<version>` or `docker.elastic.co/package-registry/distribution:production`), they will only need to update the Docker image to the latest version and restart the container to get OOTB updates for prebuilt rules.
- If a user runs EPR manually, they will need to make sure that it hosts the `security_detection_engine` package. To continue getting OOTB updates for prebuilt rules, the user would need to keep the package up to date, i.e. from time to time download the latest version of the package from https://epr.elastic.co/ and put it in the EPR's directory with packages.

**Comment by banderror** [comment]
Hi @nastasha-solomon, can we please clarify in the Rule Management docs how users of air-gapped environments could continue getting OOTB updates for prebuilt rules? @pborgonovi has done some extensive testing and you can find all the details in the thread above.

We could mention it under [Update Elastic prebuilt rules](https://www.elastic.co/docs/solutions/security/detect-and-alert/install-manage-elastic-prebuilt-rules#update-prebuilt-rules) and create a dedicated sub-page similar to [Update modified and unmodified Elastic prebuilt rules](https://www.elastic.co/docs/solutions/security/detect-and-alert/prebuilt-rules-update-modified-unmodified).

Relevant EPR docs to refer to from our docs:

- https://www.elastic.co/docs/reference/fleet/air-gapped#air-gapped-diy-epr
- https://www.elastic.co/docs/deploy-manage/deploy/self-managed/air-gapped-install#air-gapped-elastic-package-registry

**Comment by banderror** [comment]
Hey @nastasha-solomon, it's been a few months since we worked on this one with @pborgonovi. Has there been any progress on the docs side? We'd be interested in prioritizing this work if you have capacity.

**Comment by nastasha-solomon** [comment]
Hi @banderror - I should have the bandwidth to revisit this task in my upcoming sprint, which starts Nov 10, 2025. I'll set up some time with @pborgonovi during the sprint so I can get a better sense of what needs to be doc'd.

**Comment by banderror** [comment]
Thanks @nastasha-solomon! Please keep me posted when you have PRs ready for review 🙏

**Comment by akashsingh-aps** [comment]
This GitHub has so much important informations regarding upgrade process of Security prebuilt detection rules in an air-gapped env. Kudos to @banderror @pborgonovi and @nastasha-solomon . 🙏 

However I have some follow-up questions (from the discussion with customer on same topic):

- According to this the package registry container can be of a later version than the earlier ELK stack. How much later can it be? Is there any "max diff"?
- Also, how does it work with `:production` package registry (presumable latest version 9.x now) on top of an ELK stack of 8.19.5? How long will it work?

IIUC: `docker.elastic.co/package-registry/distribution:production`
- This tag is intended to always point to the latest published EPR image. When you pull and run this image, you get the most recent set of packages available in the production registry.
- If you use `:production`, you may get a newer EPR version (e.g., 8.19.8 or later) and the latest packages, depending on when you pull the image.

**Comment by banderror** [comment]
@akashsingh-aps Great questions!

> According to this the package registry container can be of a later version than the earlier ELK stack. How much later can it be? Is there any "max diff"?

I don't think Kibana is in general forward compatible with newer versions of EPR. Based on the [Fleet documentation](https://www.elastic.co/docs/reference/fleet/air-gapped#air-gapped-diy-epr), it looks like there is no guarantee for that. The recommended way is to use the versioned Docker image of the registry where the version should match the version of Kibana and other parts of the Elastic Stack:

- "_9.2.1 (recommended): docker.elastic.co/package-registry/distribution:9.2.1 - Selection of packages from the production repository released with Elastic Stack 9.2.1._"

At the top of the same [docs page](https://www.elastic.co/docs/reference/fleet/air-gapped) you can find the recommended algorithm for upgrading Stack components in an air-gapped environment.

> Also, how does it work with :production package registry (presumable latest version 9.x now) on top of an ELK stack of 8.19.5? How long will it work?

According to the above, you should use the `docker.elastic.co/package-registry/distribution:8.19.5` EPR with your 8.19.5 stack.

**Comment by akashsingh-aps** [comment]
Thank you @banderror for the pointers.

So what I understood is, we strongly recommend to use versioned [Package Registry](https://www.elastic.co/docs/reference/fleet/air-gapped#air-gapped-diy-epr) container , however if anyone willing to update the security pre-built rules (only) in an air-gapped env, they may consider using `:production` package registry to install the rules update.

Ref: https://github.com/elastic/security-docs/issues/4652#issuecomment-3056368524
https://github.com/elastic/security-docs/issues/4652#issuecomment-3114299344

Please correct me if I am still missing anything.

**Comment by banderror** [comment]
@akashsingh-aps This is only partially correct. You're missing the fact that if you use the `:production` Docker image of EPR with an old version of the stack, your Kibana may not be able to work with Fleet packages in general and install/upgrade prebuilt rules in particular, if there's a breaking change in the `:production` image from the perspective of your stack version. Does that make sense?

This is why, as I pointed in another thread, if you run a production air-gapped environment, and you want to rely on the `:production` image to have the most recent Fleet packages and prebuilt detection rules, you'd want to keep your stack always up to date (upgrade as soon as the new stack version is released) in order to minimize the chance of facing a breaking change in EPR. This may or may not work well for you in your specific case.

To summarize, the only recommended options are:

- Using the versioned EPR images, such as `docker.elastic.co/package-registry/distribution:9.2.1`, as per the [Fleet documentation](https://www.elastic.co/docs/reference/fleet/air-gapped#air-gapped-diy-epr).
- Using the `docker.elastic.co/package-registry/distribution:production` image, but only if you keep your air-gapped stack up-to-date.

**Comment by akashsingh-aps** [comment]
Thank you for clarifying this @banderror . The below two comments creates the confusion though. 

https://github.com/elastic/security-docs/issues/4652#issuecomment-3056368524
https://github.com/elastic/security-docs/issues/4652#issuecomment-3114299344

Additionally Elastic SIEM rulesets are updated more frequently than the ELK stack versions, what is the way forward in that case ?(considering the stack version is not latest). 

i.e. for the deployments running on older stack versions (for ex `v8.18.x`), how security pre-built rules can be updated ?

**Comment by banderror** [comment]
> i.e. for the deployments running on older stack versions (for ex `v8.18.x`), how security pre-built rules can be updated ?

@akashsingh-aps You can't be on an old stack version (e.g. `v8.18.x`) and have the most recent prebuilt rules that are being developed for the most recent version of the stack. You'd have to upgrade your stack to the most recent version to get the most recent rules designed for that stack. Does that make sense, conceptually?

What you can do (and this is the recommended way) is when you upgrade your air-gapped environment from, let's say, `v8.18.5` to `v8.18.6`, you upgrade your air-gapped EPR to `docker.elastic.co/package-registry/distribution:8.18.6` as well. You would get prebuilt rule updates with that, if they were shipped by Elastic in a newer version of the package with prebuilt rules created specifically for the `8.18` stack.

What you can also do (technically, at your own risk) is when you upgrade your air-gapped environment from, let's say, `v8.18.5` to `v8.18.6`, you upgrade your air-gapped EPR to `docker.elastic.co/package-registry/distribution:production`. In this case, despite the fact that you have the latest version of EPR, you wouldn't be able to get those prebuilt rule updates that were shipped for the newer stack versions. You'd still get only the updates shipped for `8.18`. Also, in this case you would increase the risk of breaking your environment if there's a breaking change in this most recent version of EPR from the `:production` image. This risk gets higher and higher the bigger becomes the gap between your stack version and the latest stack version released by Elastic.

So, we're getting back to the only [two recommended ways](https://github.com/elastic/security-docs/issues/4652#issuecomment-3540984252) to get prebuilt rule updates in an air-gapped environment. Any other ways are either not supported or not recommended.

If this feels limited for certain use cases of the users you know about, please open an enhancement request, and we'll review it separately.

**Comment by banderror** [comment]
Hey @nastasha-solomon, any progress?

**Comment by bmorelli25** [comment]
Thanks for all of the insight and context everyone! We have this issue prioritized on our board and will get to it as soon as we can.

**Comment by nastasha-solomon** [comment]
hey @banderror and @pborgonovi! A draft is ready for review in https://github.com/elastic/docs-content/pull/4972 whenever you have time.

**Comment by pborgonovi** [comment]
Hey @nastasha-solomon Thank you. I have submitted my review.

**Comment by nastasha-solomon** [comment]
Merged https://github.com/elastic/docs-content/pull/4972.
